src/main/java/org/opensearch/neuralsearch/sparse/query/PostingWithClustersScorer.java

```diff
@@ -193,9 +197,52 @@ public float score() throws IOException {
     class SingleScorer extends Scorer {
         private IteratorWrapper<DocumentCluster> clusterIter;
         private DocFreqIterator docs = null;
+        private List<DocumentCluster> allClusters;
+        private List<Float> clusterScores;
+        private int currentClusterIndex;
+        private boolean isClusterOrdered;
+        private float[] queryDenseVector;
 
-        public SingleScorer(SparsePostingsEnum postingsEnum, BytesRef term) throws IOException {
+        public SingleScorer(SparsePostingsEnum postingsEnum, BytesRef term, float[] queryDenseVector, boolean isClusterOrdered)
+            throws IOException {
             clusterIter = postingsEnum.clusterIterator();
+            this.isClusterOrdered = isClusterOrdered;
+            this.currentClusterIndex = 0;
+            this.queryDenseVector = queryDenseVector;
+            initializeClusters(postingsEnum);
+        }
+
+        private void initializeClusters(SparsePostingsEnum postingsEnum) {
+            List<DocumentCluster> clusters = new ArrayList<>();
+            List<Float> scores = new ArrayList<>();
+
+            IteratorWrapper<DocumentCluster> tempIter = clusterIter;
+            DocumentCluster cluster;
+
+            while ((cluster = tempIter.next()) != null) {
+                if (cluster.size() > 0) {
+                    float clusterScore = 0.0f;
+                    if (cluster.getSummary() != null) {
+                        clusterScore = cluster.getSummary().dotProduct(this.queryDenseVector);
+                    }
+                    clusters.add(cluster);
+                    scores.add(clusterScore);
+                }
+            }
+
+            if (this.isClusterOrdered && !clusters.isEmpty()) {
+                List<Integer> indices = IntStream.range(0, clusters.size())
+                    .boxed()
+                    .sorted((i, j) -> Float.compare(scores.get(j), scores.get(i)))
+                    .collect(Collectors.toList());
+
+                this.allClusters = indices.stream().map(clusters::get).collect(Collectors.toList());
+                this.clusterScores = indices.stream().map(scores::get).collect(Collectors.toList());
+            } else {
+                this.allClusters = clusters;
+                this.clusterScores = scores;
+            }
+            this.clusterIter = postingsEnum.clusterIterator();
```

Comment:
Since you already use allClusters as data source, there's no need to keep this clusterIter, right? You can use a new iterator for the new cluster type